### ISSUE TITLE
Pooled connections?

### ISSUE THREAD

**Issue by pthatcherg** [w3c/webrtc-quic#105]
I've recently heard interest in allowing the client/server QuicTransport to use the same QUIC connection(s) as the ones that are used for HTTP, perhaps even the one the page has used to load.  

It might work something like this:

let quic = QuicTransport.getOrCreateWithPooledConnection(hostname, port);

If there's already a QUIC connection (either QuicTransport or HTTP) to that (hostname, port), then use the same QUIC connection for the QuicTransport (and if such a QuicTransport already exists, just return the same one).  If not, create one.  If created, it can be used by future HTTP traffic as well.

You may realize this has one big problem: how do you differentiate between HTTP streams and non-HTTP streams?

Here are some options:

A.  Support datagrams.  It's not as useful has having streams, but it's still useful.  

B.  Support unidirectional streams with the new 1-byte HTTP stream type (see https://tools.ietf.org/html/draft-ietf-quic-http-17#section-3.1).  We can pick a value (only 2 out of 256 are allocated so far) and use that to indicate which streams are which.  It eats up a byte in each stream, but we can hide that from the client API, and it's still pretty useful.  

C.  Support bidirectional streams created by the server.  HTTP doesn't use them, so we could assume they aren't HTTP.  It runs a foul of the text which says 'clients MUST omit or specify a value of zero for the QUIC transport parameter "initial_max_bidi_streams"' but maybe we can work around that.

D.  Some combination of A+B+C, potentially having everything work except client-initiated bidirectional streams.

E.  Add an extension to QUIC stream frames that add a "stream group ID" allowing many protocols to use all the stream IDs but within its own group.  Kind of like many QuicTransports within one QUIC connection.  Could also be useful for running separate HTTP sessions to different hostnames which could resolve to the same host and QUIC connection.

Currently, I'm inclined toward A + B and just saying that we don't support bidirectional streams with pooled connections.  E would be nice, but it's probably too late for it to happen.

**Comment by pthatcherg** [comment]
Editor's call comments: this is interesting and probably be popular, but it would be nice if the QUIC WG didn't use all the stream IDs for HTTP.  Maybe add a stream type on bidirectional streams as well?

**Comment by lgrahl** [comment]
Sounds to me as if QUIC could use an equivalent of SCTP's PPID in addition to the stream IDs... before it's too late to introduce this change.

**Comment by pthatcherg** [comment]
Yeah, it's funny that H3 (on top of QUIC) basically added the equivalent of the PPID to unidirectional streams.  It seems like an optional uint stuffed somewhere in a QUIC stream frame would work better.

My fear is that it already is too late, but perhaps if we treated it like a generalized form of the H3 stream types applicable to all QUIC streams it might have a better chance.

**Comment by pthatcherg** [comment]
Editor's call notes: I volunteered to send an email to the QUIC WG seeing if anyone else is interested in multiplexing protocols on QUIC (HTTP + X or even HTTP + HTTP).

**Comment by vasilvv** [comment]
My current intuition would be that we should pick up A and B.  C seems too hard.  Fortunately, with WebTransport refactor, this should be easy -- just specify that pooled transport would provide one interface but not the other.

**Comment by pthatcherg** [comment]
I'll take a shot at defining this using the different mixins of WebTransport

**Comment by LPardue** [comment]
Note that the unidirectional stream type is changing from 1 byte to a variable length integer that encodes a value up to 2^62-1.
